File: python_mocra/common/models/meta.py
from typing import Dict, Any, Optional, Union
from pydantic import BaseModel, Field
from .login_info import LoginInfo

class MetaData(BaseModel):
    task: Dict[str, Any] = Field(default_factory=dict)
    login_info: Dict[str, Any] = Field(default_factory=dict)
    module_config: Dict[str, Any] = Field(default_factory=dict)
    trait_meta: Dict[str, Any] = Field(default_factory=dict)

    def add(self, key: str, value: Any, source: str) -> "MetaData":
        if source == "task":
            self.task[key] = value
        elif source == "login_info":
            self.login_info[key] = value
        elif source == "module_config":
            self.module_config[key] = value
        elif source == "trait_meta":
            self.trait_meta[key] = value
        return self

    def add_task_config(self, task_meta: Dict[str, Any]) -> "MetaData":
        # Rust replaces the whole task value if serialized
        # Here we mimic Rust's behavior: "self.task = value"
        self.task = task_meta
        return self

    def add_login_info(self, login_info: LoginInfo) -> "MetaData":
        self.login_info = login_info.extra
        return self

    def add_module_config(self, module_config: Any) -> "MetaData":
        # Assumes module_config is a dict or model with dict
        if hasattr(module_config, "model_dump"):
            self.module_config = module_config.model_dump()
        elif isinstance(module_config, dict):
            self.module_config = module_config
        return self

    def add_trait_config(self, key: str, value: Any) -> "MetaData":
        self.trait_meta[key] = value
        return self

    def get_trait_config(self, key: str) -> Optional[Any]:
        return self.trait_meta.get(key)

    def get_login_config(self, key: str) -> Optional[Any]:
        return self.login_info.get(key)

    def get_module_config(self, key: str) -> Optional[Any]:
        return self.module_config.get(key)

    def get(self, key: str, default: Any = None) -> Any:
        # Priority: trait_meta -> module_config -> task -> login_info
        if key in self.trait_meta:
            return self.trait_meta[key]
        if key in self.module_config:
            return self.module_config[key]
        if key in self.task:
            return self.task[key]
        if key in self.login_info:
            return self.login_info[key]
        return default

    def __getitem__(self, key: str) -> Any:
        val = self.get(key)
        if val is None:
            raise KeyError(key)
        return val

    def __setitem__(self, key: str, value: Any):
        self.trait_meta[key] = value

File: crates/mocra-core/src/common/model/meta.rs
use crate::common::model::ModuleConfig;
use crate::common::model::login_info::LoginInfo;

use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Default, Debug, Clone, Serialize, Deserialize)]
pub struct MetaData {
    /// Task metadata. Provenance: Task definition → ParserModel/ErrorModel processing →
    /// TaskFactory merge → Module.generate() → Request.meta
    /// Stored by serializing the task config into a Value via `add_task_config()`.
    #[serde(with = "crate::common::model::serde_value::value")]
    pub task: Value,
    /// Login information. Provenance: the LoginInfo.extra field → user authentication/authorization
    /// system → Module.generate() → Request.meta
    /// Extracted from LoginInfo's extra field and serialized via `add_login_info()`.
    #[serde(with = "crate::common::model::serde_value::value")]
    pub login_info: Value,
    /// Module configuration. Provenance: database (account/platform/middleware relation tables) →
    /// ModuleConfig instance → module initialization → Module.generate() → Request.meta
    /// Includes: account table config, platform table config, the module's own config,
    /// rel_module_account config, rel_module_platform config, middleware config, and so on.
    /// Stored by serializing the whole ModuleConfig struct via `add_module_config()`.
    #[serde(with = "crate::common::model::serde_value::value")]
    pub module_config: Value,
    /// Trait metadata. Provenance: individual trait implementations → trait type definitions →
    /// business code calls → add_trait_config() → Request.meta
    /// Stores trait-level custom configuration and extension data via `add_trait_config()`.
    #[serde(with = "crate::common::model::serde_value::value")]
    pub trait_meta: Value,
}

/// Type-safe selector for MetaData storage slots.
#[derive(Debug, Clone, Copy)]
pub enum MetaSource {
    Task,
    LoginInfo,
    ModuleConfig,
    TraitMeta,
}

impl MetaData {
    pub fn add(
        mut self,
        key: impl AsRef<str>,
        value: serde_json::Value,
        source: MetaSource,
    ) -> Self {
        let target = match source {
            MetaSource::Task => &mut self.task,
            MetaSource::LoginInfo => &mut self.login_info,
            MetaSource::ModuleConfig => &mut self.module_config,
            MetaSource::TraitMeta => &mut self.trait_meta,
        };

        if !target.is_object() {
            *target = Value::Object(serde_json::Map::new());
        }

        if let Some(map) = target.as_object_mut() {
            map.insert(key.as_ref().into(), value);
        }
        self
    }
    pub fn add_task_config<T>(mut self, task_meta: T) -> Self
    where
        T: Serialize,
    {
        if let Ok(value) = serde_json::to_value(task_meta) {
            self.task = value;
        }
        self
    }
    pub fn add_login_info(mut self, login_info: &LoginInfo) -> Self {
        if let Ok(value) = serde_json::to_value(&login_info.extra) {
            self.login_info = value;
        }
        self
    }
    pub fn add_module_config(mut self, module_config: &ModuleConfig) -> Self {
        if let Ok(value) = serde_json::to_value(module_config) {
            self.module_config = value;
        };

        self
    }
    pub fn add_trait_config<T>(mut self, key: impl AsRef<str>, value: T) -> Self
    where
        T: Serialize,
    {
        if !self.trait_meta.is_object() {
            self.trait_meta = Value::Object(serde_json::Map::new());
        }

        if let Some(map) = self.trait_meta.as_object_mut() {
            if let Ok(value) = serde_json::to_value(value) {
                map.insert(key.as_ref().into(), value);
            }
        }

        self
    }
    pub fn get_trait_config<T>(&self, key: &str) -> Option<T>
    where
        T: for<'de> Deserialize<'de>,
    {
        self.trait_meta
            .get(key)
            .cloned()
            .and_then(|v| serde_json::from_value(v).ok())
    }
    pub fn get_login_config<T>(&self, key: &str) -> Option<T>
    where
        T: for<'de> Deserialize<'de>,
    {
        self.login_info
            .get(key)
            .cloned()
            .and_then(|v| serde_json::from_value(v).ok())
    }
    pub fn get_module_config<T>(&self, key: &str) -> Option<T>
    where
        T: for<'de> Deserialize<'de>,
    {
        self.module_config
            .get(key)
            .cloned()
            .and_then(|v| serde_json::from_value(v).ok())
    }
    pub fn get_task_config<T>(&self, key: &str) -> Option<T>
    where
        T: for<'de> Deserialize<'de>,
    {
        self.task
            .get(key)
            .cloned()
            .and_then(|v| serde_json::from_value(v).ok())
    }
}

impl From<MetaData> for Value {
    fn from(value: MetaData) -> Self {
        serde_json::to_value(value).unwrap_or_default()
    }
}
